src/fmu/dataio/_fmu_provider.py

```diff
@@ -1,292 +1,387 @@
-    case_name: Optional[str] = field(default=None, init=False)
-    user_name: Optional[str] = field(default=None, init=False)
-    ert2: dict = field(default_factory=dict, init=False)
-    case_metafile: Optional[Path] = field(default=None, init=False)
-    case_metadata: dict = field(default_factory=dict, init=False)
-    metadata: dict = field(default_factory=dict, init=False)
-    rootpath: Optional[Path] = field(default=None, init=False)
-
-    def __post_init__(self) -> None:
-        self.rootpath = Path(self.dataio._rootpath.absolute())
+    """Class for detecting the run environment (e.g. an ERT) and provide metadata.
+
+    Args:
+        model: Name of the model (usually from global config)
+        rootpath: ....
+        fmu_context: The FMU context this is ran in; see FmuContext enum class
+        casepath_proposed: Proposed casepath ... needed?
+        include_ertjobs: True if we want to include ....
+        forced_realization: If we want to force the realization (use case?)
+        workflow: Descriptive work flow info
+    """
 
-        self.rootpath_initial = self.rootpath
+    model: str = ""
+    fmu_context: FmuContext = FmuContext.REALIZATION
+    include_ertjobs: bool = True
+    casepath_proposed: str | Path = ""
+    forced_realization: Optional[int] = None
+    workflow: str | dict = ""
+
+    # private properties for this class
+    _stage: str = field(default="unset", init=False)
+    _runpath: Optional[Path] = field(default=None, init=False)
+    _casepath: Optional[Path] = field(default=None, init=False)  # actual casepath
+    _provider: Optional[str] = field(default=None, init=False)
+    _iter_name: Optional[str] = field(default=None, init=False)
+    _iter_id: Optional[int] = field(default=None, init=False)
+    _iter_path: Optional[Path] = field(default=None, init=False)
+    _real_name: Optional[str] = field(default=None, init=False)
+    _real_id: int = field(default=0, init=False)
+    _real_path: Optional[Path] = field(default=None, init=False)
+    _case_name: Optional[str] = field(default=None, init=False)
+    _user_name: Optional[str] = field(default=None, init=False)
+    _ert_info: dict = field(default_factory=dict, init=False)
+    _case_metadata: dict = field(default_factory=dict, init=False)
+    _metadata: dict = field(default_factory=dict, init=False)
 
-        logger.info("Initialize %s", self.__class__)
+    def __post_init__(self) -> None:
+        logger.debug("Initialize %s...", self.__class__)
+        logger.debug("Case path is initially <%s>...", self.casepath_proposed)
+        logger.debug("FMU context is <%s>...", self.fmu_context)
+
+        if not FmuEnv.ENSEMBLE_ID.value:
+            return  # not an FMU run
+
+        self._provider = "ERT"
+
+        self._detect_fmurun_stage()
+        self._detect_absolute_runpath()
+        self._detect_and_update_casepath()
+        self._parse_folder_info()
+        self._read_case_metadata()
+
```

Comment:
I know Im hard on the pydantic juice, but i cant stop to wondering if these would read better by using decorator model_validator(mode="after"). It would also protect well against users trying to work around any the setup logic implemented in the __post_init__(...).